torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3275,39 +3428,62 @@ def evaluate_expr(self, orig_expr: "sympy.Expr", hint=None, fx_node=None):
             guard = ShapeGuard(g, stack)
             self.guards.append(guard)
             self.refine_ranges(guard)
-            if self.log.isEnabledFor(logging.INFO):
-                for frame in reversed(tb):
-                    if frame.filename not in uninteresting_files():
-                        break
-
-                # NB: this stack is truncated, but it's fine because the main
-                # stack_info will give you the rest of the info you need
-                maybe_user_loc = ""
-                user_tb = TracingContext.extract_stack()
-                if user_tb:
-                    maybe_user_loc = " at " + format_frame(user_tb[-1])
-
-                is_debug = self.log.isEnabledFor(logging.DEBUG)
-                maybe_extra_debug = ""
-                if is_debug and user_tb:
-                    maybe_extra_debug = (
-                        '\nUser Stack (most recent call last):\n' +
-                        '  (snipped, see stack below for prefix)\n' +
-                        ''.join(traceback.format_list(user_tb))
-                    )
-                self.log.info(
-                    "eval %s [guard added]%s (%s)%s",
-                    g,
-                    maybe_user_loc,
-                    format_frame(frame),
-                    maybe_extra_debug,
-                    stack_info=is_debug,
-                )
+            self._log_guard("eval", g, tb)
         else:
             self.log.debug("eval %s [guard suppressed]", g)
 
         return concrete_val
 
+    def defer_runtime_assert(self, orig_expr: "sympy.Expr", msg, fx_node=None):
+        expr = orig_expr
+
+        static_expr = self._maybe_evaluate_static(expr)
+        if static_expr is not None:
+            self.log.debug("runtime_assert %s == %s [statically known]", orig_expr, static_expr)
+            return static_expr
+
+        # Attempt to eliminate the unbacked SymInt
+        new_expr = self._maybe_evaluate_static(expr, unbacked_only=True)
+        if new_expr.free_symbols <= self.var_to_val.keys():
+            # Do a normal guard
+            return self.evaluate_expr(new_expr, fx_node=fx_node)
+        # NB: Don't use new_expr as expr; it could contain gunk like shape0
+        # which we don't want to guard on
+
+        # OK, we're definitely doing a runtime assert now
+        if (
+            _translation_validation_enabled()
+            and fx_node is not None
+            and not self._suppress_guards_tls()
+        ):
+            self.create_fx_call_function(torch._assert, (fx_node,))
```

Comment:
If we make it a **source** expression, then there must be a corresponding **target** expression somewhere. Since these runtime assertions are supposed to be always true, shouldn't it be simply a TV assertion or **target** expression?

Reply:
Ah yes, good point. I think we need to "generate" these target expressions during produce guards.